### PR TITLE
[RLLib] Add additional return values to action_sampler_fn.

### LINKED ISSUE #21648: Custom action_sampler_fn is not working for PPO.
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

RLlib

### What happened + What you expected to happen

PPO does not work while using `action_sampler_fn` and `make_model`.

```
 File "/lib/python3.8/site-packages/ray/rllib/evaluation/rollout_worker.py", line 587, in __init__
    self._build_policy_map(
  File "/lib/python3.8/site-packages/ray/rllib/evaluation/rollout_worker.py", line 1551, in _build_policy_map
    self.policy_map.create_policy(name, orig_cls, obs_space, act_space,
  File "/lib/python3.8/site-packages/ray/rllib/policy/policy_map.py", line 133, in create_policy
    self[policy_id] = class_(
  File "/lib/python3.8/site-packages/ray/rllib/policy/tf_policy_template.py", line 238, in __init__
    DynamicTFPolicy.__init__(
  File "/lib/python3.8/site-packages/ray/rllib/policy/dynamic_tf_policy.py", line 376, in __init__
    self._initialize_loss_from_dummy_batch(
  File "/lib/python3.8/site-packages/ray/rllib/policy/dynamic_tf_policy.py", line 649, in _initialize_loss_from_dummy_batch
    losses = self._do_loss_init(train_batch)
  File "/lib/python3.8/site-packages/ray/rllib/policy/dynamic_tf_policy.py", line 731, in _do_loss_init
    losses = self._loss_fn(self, self.model, self.dist_class, train_batch)
  File "/python3.8/site-packages/ray/rllib/agents/ppo/ppo_tf_policy.py", line 56, in ppo_surrogate_loss
    curr_action_dist = dist_class(logits, model)
TypeError: 'NoneType' object is not callable
```
### Versions / Dependencies

Python =3.8
ray==1.9.2

### Reproduction script

```
from typing import Any, Optional, Tuple, Type

import numpy as np
import ray
from gym.spaces import Box, Space
from ray.rllib.agents.ppo import ppo, ppo_tf_policy
from ray.rllib.agents.trainer_template import build_trainer
from ray.rllib.models.catalog import ModelCatalog
from ray.rllib.models.modelv2 import ModelV2
from ray.rllib.policy.policy import Policy
from ray.rllib.policy.sample_batch import SampleBatch
from ray.rllib.policy.tf_policy_template import build_tf_policy
from ray.rllib.utils.framework import try_import_tf
from ray.rllib.utils.spaces.simplex import Simplex
from ray.rllib.utils.tf_utils import zero_logps_from_actions
from ray.rllib.utils.typing import TensorType, TrainerConfigDict

tf1, tf, tfv = try_import_tf()


def make_model(
    policy: Policy, obs_space: Space, action_space: Space, config: TrainerConfigDict
) -> ModelV2:
    if isinstance(action_space, Box):
        num_outputs = 2 * np.product(action_space.shape)
    model = ModelCatalog.get_model_v2(
        obs_space=obs_space,
        action_space=action_space,
        num_outputs=num_outputs,
        model_config=config["model"],
        framework="tf",
    )
    policy.dist_class, _ = ModelCatalog.get_action_dist(action_space, config["model"])
    return model


def action_sampler_fn(
    policy: Policy,
    model: ModelV2,
    obs_batch: TensorType,
    explore: bool = True,
    state_batches: Optional[TensorType] = None,
    seq_lens: Optional[TensorType] = None,
    prev_action_batch: Optional[TensorType] = None,
    prev_reward_batch: Optional[TensorType] = None,
    **kwargs: Any,
) -> Tuple[TensorType, TensorType]:
    distribution_inputs, policy._state_out = policy.model(
        {
            SampleBatch.OBS: obs_batch,
            "obs_flat": obs_batch,
            "is_training": policy._get_is_training_placeholder(),
            SampleBatch.PREV_ACTIONS: prev_action_batch,
            SampleBatch.PREV_REWARDS: prev_reward_batch,
        },
        state_batches,
        seq_lens,
    )
    action_dist_class, _ = ModelCatalog.get_action_dist(policy.action_space, policy.config["model"])
    action_dist = action_dist_class(distribution_inputs, model)
    action = action_dist.deterministic_sample()

    logp = zero_logps_from_actions(action)
    return action, logp


PPOTFPolicy = build_tf_policy(
    name="PPOTFPolicy",
    loss_fn=ppo_tf_policy.ppo_surrogate_loss,
    make_model=make_model,
    action_sampler_fn=action_sampler_fn,
    get_default_config=lambda: ray.rllib.agents.ppo.ppo.DEFAULT_CONFIG,
    postprocess_fn=ppo_tf_policy.compute_gae_for_sample_batch,
    stats_fn=ppo_tf_policy.kl_and_loss_stats,
    compute_gradients_fn=ppo_tf_policy.compute_and_clip_gradients,
    extra_action_out_fn=ppo_tf_policy.vf_preds_fetches,
    before_init=ppo_tf_policy.setup_config,
    before_loss_init=ppo_tf_policy.setup_mixins,
    mixins=[
        ppo_tf_policy.LearningRateSchedule,
        ppo_tf_policy.EntropyCoeffSchedule,
        ppo_tf_policy.KLCoeffMixin,
        ppo_tf_policy.ValueNetworkMixin,
    ],
)

DEFAULT_CONFIG = ppo.DEFAULT_CONFIG


PPOTrainer = build_trainer(
    name="PPO",
    default_config=ppo.DEFAULT_CONFIG,
    validate_config=ppo.validate_config,
    default_policy=PPOTFPolicy,
    execution_plan=ppo.execution_plan,
)


config = DEFAULT_CONFIG.copy()
trainer = PPOTrainer(config, env="LunarLanderContinuous-v2")
for i in range(250):
    trainer.train()
```
### Anything else

_No response_

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by n30111**
## Why are these changes needed?

Currently PPO does not work while using `action_sampler_fn` and `make_model`. Adding these additional returns values to `action_sampler_fn` solve the issue.

## Related issue number

[Closes #21648](https://github.com/ray-project/ray/issues/21648)

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by n30111** [comment]
@gjoliver thanks for your comments, I have updated the code and related doc string.

**Review by gjoliver — COMMENTED** [review]
Really appreciate the work Ishant.
I looked at our code base, and realize there aren't that many custom action_sampler_fn we need to update, maybe 2.
right now, TF version of action_sampler_fn gives 2 outputs, and Torch version already gives 3 (including states).
I am thinking maybe we should just update everything and standardize on 4 return values.
let me loop in @sven1977 and see what he thinks.

File: rllib/policy/dynamic_tf_policy.py [hunk 1]
```diff
@@ -303,6 +310,17 @@ def __init__(
                     explore=explore,
                     is_training=self._input_dict.is_training,
                 )
+                if len(action_sampler_outputs) == 4:
+                    (
+                        sampled_action,
+                        sampled_action_logp,
+                        dist_inputs,
```
Comment (avnishn) [inline rllib/policy/dynamic_tf_policy.py:317]: should we not include this if statement and instead generically return the action sampler outputs? Users may want to expand the info that is returned by the action sampler -- I can't think of the usecase, but its possible CC @gjoliver

**Review by avnishn — COMMENTED** [review]
Left a comment about whether we should assume 2 or 4 returns 

maybe what we could do is instead expect that action sampling fn returns a dictionary, that way the returns are named and therefore unambigous

File: rllib/policy/dynamic_tf_policy.py (hunk 1 above)
Comment (gjoliver) [reply rllib/policy/dynamic_tf_policy.py:317]: that may be a bigger change compatibility-wise?
these data fields are actually immediately used below.

**Review by sven1977 — APPROVED** [review]
Looks good to me. Thanks for this PR @n30111 .
+1 merging it as RLlib itself almost never (except Dreamer) uses `action_sampler_fn` and this actually would allow users to also customize action sampling behavior + RNNs.

**Comment by n30111** [comment]
@gjoliver , is any action required my side on this PR?

**Comment by gjoliver** [comment]
> @gjoliver , is any action required my side on this PR?

I don't know if you want to help standardize this to 4 return values all the time, which basically means that we don't do this output then "if len()" check, and simply always try to unpack 4 values for both tf and torch calls.
We will then also update the couple of current RLlib users of action_sampler_fn to return 4 values.
I understand this may be more work than you would like. So if you just want to merge current version, that works too, and we can deprecate later.
thanks for the helps again.

**Comment by gjoliver** [comment]
@sven1977 can you help merge please? thanks

**Pull request merged**